### DISCUSSION TITLE
Is Docker Label Handling a nginx-proxy-manager Feature?

### DISCUSSION THREAD

**Discussion by Sascha-Block** [NginxProxyManager/nginx-proxy-manager#1268] [Q&A]
Labels are a mechanism for applying metadata to Docker objects, including:

- Images
- Containers
- Local daemons
- Volumes
- Networks
- Swarm nodes
- Swarm services

You can use labels to organize your images, record licensing information, annotate relationships between containers, volumes, and networks, or in any way that makes sense for your business or application.

Does nginx-proxy-manager provide Docker Label Handling as a Feature?

**Comment by chaptergy** [comment]
Short answer: no.

In what ways could it use those labels? Grouping the hosts? Suggesting new hosts? Is it even possible to get access to this label or labels of other containers from inside a docker container?

**Reply by Sascha-Block** [reply]
The labels can be used to define the routing. This is done directly in the yaml file and is quite intuitive.

A use case would be **service injection**. 

Example:

If a patch is applied on a server and there is a downtime in connection with this, a page with service information can be comfortably displayed in the front end. 

Other tools eg. Traefik allows to simply forward requests to a defined error page container.

Loadbalacing is also made very simple. Everything is done with 1-3 lines and the tracing feature is great. So you can see "which way" the requests take in your own infrastructure and make adjustments if necessary.

Such an easy handling would be a wish within the nginx-proxy-manager project...

**Reply by chaptergy** [reply]
I'm not sure I would know enough about the topic to be able to implement it. This most likely would have to be contributed by someone from the community. But automatic routing is always nice.
Somewhat related, someone created a separate utility in https://github.com/jc21/nginx-proxy-manager/issues/506#issuecomment-660952688, which automatically creates hosts based on environment variables on the containers.

**Comment by tarach** [comment]
I had the same need: I wanted to keep using Nginx Proxy Manager, but define proxy hosts through Docker labels instead of creating them manually in the UI every time.

So I built a small companion container for NPM:

https://github.com/tarach/npm-docker-auto-proxy

It watches Docker container events and uses the NPM API to create, update, enable, disable, or delete proxy hosts based on container labels.

It does not replace Nginx Proxy Manager. NPM still handles the reverse proxy and remains the UI/source of truth — this only adds a Docker-label automation layer around it.

Example:

```yaml
labels:
  npm.proxy.enabled: "true"
  npm.proxy.domain: "jellyfin.example.com"
  npm.proxy.forward_host: "jellyfin"
  npm.proxy.forward_port: "8096"
  npm.proxy.scheme: "http"
  npm.proxy.ssl: "true"
  npm.proxy.certificate: "*.example.com"
  npm.proxy.force_ssl: "true"
```

Just to be clear: I did not create this as competition to projects like:

* https://github.com/Xantios/nginx-proxy-manager-assistant
* https://github.com/Redth/npm-docker-sync

I simply could not find them earlier when I started working on my own helper. 😅  
I hope there is still some value in another implementation.

Maybe this helps others who like NPM but miss Traefik-like Docker labels.
